### ISSUE TITLE
ThunderSVM accuracy degraded for large C parameter

### ISSUE THREAD

**Issue by wirawan0** [Xtra-Computing/thundersvm#125]
Attached please find a testcase (input/output files and code) for a case where ThunderSVM (using Python scikit-learn interface) shows divergence from the standard LibSVM (i.e. the original `sklearn.svm.SVC` implementation). When the C parameter is small enough (up to 1000 for the testcase enclosed) both show the same result (statistically). But when C grows larger (to 10,000 or even 100,000) then the accuracy of the SVM produced by ThunderSVM drops to 65% then to 53%. But it is odd that I found out (by turning on the "Verbose" option) that the rho and nSV parameters are about the same between the classifier returned by ThunderSVM and libSVM. Can someone explain what the problem is?

It looks to me that the converger used by ThunderSVM has a different behavior under extreme C parameters. But I don't know how to adjust its behavior.

Here are representative outputs for quick preview: produced using `grep -e '----' -e 'Accuracy' -e 'Fitting' debug_svm1.out` command:

    ---- Standard parameters ----
    Fitting model1 (ThunderSVM)
     Accuracy score (for testing data): 0.81
    Fitting model2 (LibSVM)
    Accuracy score (for testing data): 0.81
    ---- Tweaked parameters C=100000.0, gamma=0.001 ----
    Fitting model3 (ThunderSVM)
    Accuracy score (for testing data): 0.5316666666666666
    Fitting model5 (ThunderSVM)
    Accuracy score (for testing data): 0.5316666666666666
    Fitting model4 (LibSVM)
    Accuracy score (for testing data): 0.8183333333333334
    ---- Tweaked parameters C=10000.0, gamma=0.001 ----
    Fitting model7 (ThunderSVM)
    Accuracy score (for testing data): 0.6516666666666666
    Fitting model8 (LibSVM)
    Accuracy score (for testing data): 0.8183333333333334
    ---- Tweaked parameters C=1000.0, gamma=0.001 ----
    Fitting model9 (ThunderSVM)
    Accuracy score (for testing data): 0.815
    Fitting model10 (LibSVM)
    Accuracy score (for testing data): 0.81

**Comment by wirawan0** [comment]
Code file (python3)

**Comment by wirawan0** [comment]
Attachment included:

    -rw-r--r-- wp/users 4094 2019-02-11 00:24 debug_svm1.out
    -rwx------ wp/users 3390 2019-02-11 00:18 debug_svm1.py
    -rw-r--r-- wp/users 89255 2019-02-10 21:38 test_data.svm
    -rw-r--r-- wp/users 356612 2019-02-10 21:38 train_data.svm

**Comment by wirawan0** [comment]
Attachment included (finally):

    -rw-r--r-- wp/users 4094 2019-02-11 00:24 debug_svm1.out
    -rwx------ wp/users 3390 2019-02-11 00:18 debug_svm1.py
    -rw-r--r-- wp/users 89255 2019-02-10 21:38 test_data.svm
    -rw-r--r-- wp/users 356612 2019-02-10 21:38 train_data.svm

[debug_svm1.tar.gz](https://github.com/Xtra-Computing/thundersvm/files/2849947/debug_svm1.tar.gz)

**Comment by wirawan0** [comment]
Sorry... I need to fix the data. Looks like thundersvm expects 1-based sparse matrix in the libsvm input files? Attached are the files with the 1-based indices. Please use these instead of the *.svm files in the tar file.

[test_data.txt](https://github.com/Xtra-Computing/thundersvm/files/2852982/test_data.txt)
[train_data.txt](https://github.com/Xtra-Computing/thundersvm/files/2852983/train_data.txt)

Extensions have been changed to `.txt`. Please rename to `.svm` if using my script.

**Comment by wirawan0** [comment]
I can confirm that the dropped accuracy is also observed when testing the dataset with the command-line programs `thundersvm-train` and `thundersvm-predict`.

Build info: My thundersvm library was compiled for 64-bit Linux, CPU-only (with AVX2 capable hardware--Xeon E5 'Haswell' class CPU). I was using the latest master branch of the git repo (commit 1ab487f397321).

**Comment by zeyiwen** [comment]
Thanks for the feedback! We have reproduced the problem, and are working on fixing it.

**Comment by zeyiwen** [comment]
@wirawan0 Hi there. It appears to be a precision issue in the prediction module. We have fixed it. Please pull the latest version, and try it out. If this issue is solved, please close it. Thanks!

**Comment by zeyiwen** [comment]
The issue should be fixed now, and hence we like to close it.

**Comment by wirawan0** [comment]
Sorry for a late reponse. I tested this today (commit 6062b4f) and the ThunderSVM can now produce same result (statistically) as libSVM. Thanks for fixing the issue!
